### PR TITLE
task/JM-7989

### PR DESCRIPTION
**Pull request by akikrahman1**
### JIRA link (if applicable) ###

https://centralgovernmentcgi.atlassian.net/browse/JM-7989

### Change description ###

Updated the court deferral denied letter with queryDsl queries

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java [hunk 1]
```diff
@@ -1,83 +1,194 @@
...
+
+        // run query to get the most recent deferral denied date for each juror
+        List<Tuple> printedDates = getPrintedDate(owner, poolNumbers);
+
+        // update the deferral denied printed date for each juror
+        printedDates.forEach(tuple -> {
+            String jurorNumber = tuple.get(QJuror.juror.jurorNumber);
+            if (deferralDeniedLetterListMap.containsKey(jurorNumber)) {
+                DeferralDeniedLetterList deferralDeniedLetterList = deferralDeniedLetterListMap.get(jurorNumber);
+                LocalDateTime datePrinted = tuple.get(QJurorHistory.jurorHistory.dateCreated);
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:84]: Can you grab convert toLocalDate() here so it is done once

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java [hunk 2]
```diff
@@ -1,83 +1,194 @@
...
+
+    }
+
+
+    private JPAQuery<Tuple> buildBaseQuery(String owner) {
+
+        JPAQueryFactory queryFactory = new JPAQueryFactory(entityManager);
+
+        return queryFactory.select(
+                QJurorPool.jurorPool.pool.poolNumber,
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:107]: Should use QPoolRequest.poolRequest as we have a direct join for this

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java [hunk 3]
```diff
@@ -1,83 +1,194 @@
...
+            .from(QJurorPool.jurorPool)
+            .join(QJuror.juror).on(QJuror.juror.eq(QJurorPool.jurorPool.juror))
+            .join(QPoolRequest.poolRequest)
+            .on(QPoolRequest.poolRequest.eq(QJurorPool.jurorPool.pool))
+            .join(QJurorStatus.jurorStatus).on(QJurorStatus.jurorStatus.eq(QJurorPool.jurorPool.status))
+            .join(QJurorHistory.jurorHistory)
+            .on(QJurorHistory.jurorHistory.jurorNumber.eq(QJuror.juror.jurorNumber)
+                .and(QJurorHistory.jurorHistory.poolNumber.eq(QPoolRequest.poolRequest.poolNumber))
+                .and(QJurorHistory.jurorHistory.historyCode.eq(HistoryCodeMod.DEFERRED_POOL_MEMBER))
+                .and(QJurorHistory.jurorHistory.otherInformation.containsIgnoreCase("deferral denied"))
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:126]: would be better to just be contains (for better index performance) assuming we know the case deferral denied will be in

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java [hunk 4]
```diff
@@ -1,83 +1,194 @@
...
+                .and(QJurorHistory.jurorHistory.dateCreatedDateOnly.after(QJuror.juror.bureauTransferDate)))
+            .where(QJuror.juror.excusalRejected.eq("Z")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:128]: Z would be better to come from the Excusal codes class

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java [hunk 5]
```diff
@@ -1,83 +1,194 @@
...
+                .and(QPoolRequest.poolRequest.owner.eq(owner)));
     }
 
+    private List<Tuple> getPrintedDate(String owner, List<String> poolNumbers) {
+
+        JPAQueryFactory queryFactory = new JPAQueryFactory(entityManager);
 
-    private void filterEligibleLetterSearchCriteria(JPAQuery<DeferralDeniedLetterList> jpaQuery,
+        return queryFactory.select(
+                QJurorPool.jurorPool.pool.poolNumber,
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:138]: Same as above use QPoolRequest.poolRequest

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java [hunk 6]
```diff
@@ -1,83 +1,194 @@
...
+            .from(QJurorPool.jurorPool)
+            .join(QJuror.juror).on(QJuror.juror.eq(QJurorPool.jurorPool.juror))
+            .join(QPoolRequest.poolRequest)
+            .on(QPoolRequest.poolRequest.eq(QJurorPool.jurorPool.pool))
+            .join(QJurorStatus.jurorStatus).on(QJurorStatus.jurorStatus.eq(QJurorPool.jurorPool.status))
+            .join(QJurorHistory.jurorHistory)
+            .on(QJurorHistory.jurorHistory.jurorNumber.eq(QJuror.juror.jurorNumber)
+                .and(QJurorHistory.jurorHistory.historyCode.eq(HistoryCodeMod.NON_DEFERRED_LETTER))
+                .and(QJurorHistory.jurorHistory.dateCreatedDateOnly.after(QJuror.juror.bureauTransferDate)))
+            .where(QJuror.juror.excusalRejected.eq("Z")
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:150]: Same as above would be better to use the constant class for Z

**Review by Ben-Edwards-cgi — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java (hunk 1 above)
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:84]: done

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java (hunk 2 above)
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:107]: updated

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java (hunk 3 above)
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:126]: updated

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java (hunk 5 above)
Comment (akikrahman1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/letter/court/DeferralDeniedLetterListRepositoryImpl.java:138]: updated

**Pull request merged**